### DISCUSSION TITLE
CalendarDatePicker Styling

### DISCUSSION THREAD

**Discussion by ManuelKetisch** [AvaloniaUI/Avalonia#13854] [Q&A]
Hello,

i am trying to style a CalendarDatePicker Control. So far i only managed to style the main body of the date picker. I am looking for a way to style the parts from the previous and the coming month as well as the upper part where the month and year can be picked.

I marked the parts i want to style in the posted image.
How can i access these parts of the control?

![CalendarDatePicker_Styling](https://github.com/AvaloniaUI/Avalonia/assets/50184702/c262da09-ed9a-4c4c-9719-d017afa84571)

**Comment by thevortexcloud** [comment]
This discussion may help
https://github.com/AvaloniaUI/Avalonia/discussions/13642

**Reply by ManuelKetisch** [reply]
No sorry, that did not really clarify how i access the highlighted parts of the control.

**Reply by timunie** [reply]
easiest is to lookup original styles and override the resources or styles. If you use FluentTheme find them here: https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Themes.Fluent/Controls

Moreover DevTools [F12] can be helpful.

**Reply by ManuelKetisch** [reply]
Yes i already looked at the original styles, but still can't figure out which ones are responsible for the button styles or how to access them. What is very confusing is the fact, that parts of the calendar control buttons have a different color than the rest of the buttons.

**Reply by timunie** [reply]
That works for me for example for the out of range dates: 

```xml
<Style Selector="CalendarDayButton:inactive">
    <Style Selector="^ /template/ Border#Root">
        <Setter Property="Background" Value="Red" />
    </Style>
    <Style Selector="^ /template/ ContentControl#Content">
        <Setter Property="Foreground" Value="CornflowerBlue" />
    </Style>
</Style>
```

result: 
![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/85ee4b0d-2aa8-43f6-9113-cae2798c97ed)

original style: 
https://github.com/AvaloniaUI/Avalonia/blob/0a6987b0855bc7cd0b371571bb0234dc779da181/src/Avalonia.Themes.Fluent/Controls/CalendarDayButton.xaml#L96-L103

**Reply by ManuelKetisch** [reply]
Ok that works for me too, for the out of range dates, thank you for helping out.
I did not realize, that it was actually another control that i had to style "CalendarDayButton".
This is quite confusing. How can i figure this out on my own? 
The only parts that need styling now, is the Button / ComboBox type of control on top of the CalendarDatePicker control and the SpinnerButton to the right.

**Reply by timunie** [reply]
scan style files and also play around with the several styles to get used to the syntax. I don't do it different.

**Reply by ManuelKetisch** [reply]
Ok i managed to style the upper elements of the CalendarDatePicker.
Those where actually CalendarItems:

```
        <Style Selector="CalendarItem /template/ Button#PART_HeaderButton">
            <Setter Property="Background" Value="#175381" />
            <Setter Property="Margin" Value="-11 0 0 0"/>
        </Style>
        <Style Selector="CalendarItem /template/ Button#PART_PreviousButton">
            <Setter Property="Background" Value="#175381"/>
        </Style>
        <Style Selector="CalendarItem /template/ Button#PART_NextButton">
            <Setter Property="Background" Value="#175381" />
        </Style>
```

Now there is only one part of the Control left that can't figure out.
Whenever the HeaderButton is clicked, the controls view switches from day selection to month selection.
Another click on that Button switches it to year selection.
The Background and Foreground of these two view states are not tied to the standard day selection when you first open the control.
I can use the following styling to change the day selection of the current month:
```
        <Style Selector="CalendarItem">
            <Style Selector="^ /template/ Grid#PART_MonthView">
                <Setter Property="Background" Value="Red"/>
            </Style>
        </Style>
```
But i can not change the background of the YearView: (This does not seem to do anything.)
```
        <Style Selector="CalendarItem">
            <Style Selector="^ /template/ Grid#PART_YearView">
                <Setter Property="Background" Value="Red"/>
            </Style>
        </Style>
```

With:
```
      <Style Selector="CalendarButton:inactive">
        <Style Selector="^ /template/ Border#Root">
          <Setter Property="Background" Value="#858585"/>
        </Style>
      </Style>
```
I can only change the first and last button. 
Using CalendarButton without its pseudoclass 'inactive' like this:
```
      <Style Selector="CalendarButton">
        <Style Selector="^ /template/ Border#Border">
          <Setter Property="Background" Value="Red"/>
        </Style>
      </Style>
```
changes the color of all buttons, but the foreground disappears.
Any idea how i can change the Background and Foreground of that views buttons?
Below is a screenshot of the control i want to style:
![image](https://github.com/AvaloniaUI/Avalonia/assets/50184702/7b3997a0-4a7e-4eae-92d6-30cae64b4bdb)

**Comment by arkadym** [comment]
Not sure if yet actual, i finally managed to style CalendarDatePicker, it requries to copy CalendarItem template and littlebit change it (see attached xaml)
[CalendarDatePicker.patch](https://github.com/user-attachments/files/24769879/CalendarDatePicker.patch)
[CalendarDatePicker.xml](https://github.com/user-attachments/files/24770109/CalendarDatePicker.xml)
<img width="359" height="459" alt="image" src="https://github.com/user-attachments/assets/5c324758-ea87-4d37-ae81-6a9b20f0d6de" />
<img width="313" height="395" alt="image" src="https://github.com/user-attachments/assets/a0324fc1-3d82-4b14-b9e3-7faa49e38140" />
<img width="365" height="395" alt="image" src="https://github.com/user-attachments/assets/ee6f1763-3099-42f1-8f78-cb7f16c192c0" />

**Reply by timunie** [reply]
Beautiful design

**Reply by LaurentInSeattle** [reply]
Nice! 
Does it globalize correctly date formats and first day of week ?
